peas/tests/serial_handlers/protocol_arduino_simulator.py

```diff
@@ -0,0 +1,422 @@
+        self.message = self.telemetry_board
+        self.logger.info('ArduinoSimulator created')
+
+    def __del__(self):
+        if not self.stop.is_set():
+            self.logger.critical('ArduinoSimulator.__del__ stop is NOT set')
+
+    def run(self):
+        """Produce messages periodically and emit their bytes at a limited rate."""
+        self.logger.info('ArduinoSimulator.run ENTER')
+        # Produce a message right away, but remove a random number of bytes at the start to reflect
+        # what happens when we connect at a random time to the Arduino.
+        now = datetime.datetime.now()
+        self.next_chunk_time = now
+        self.next_message_time = now + self.message_delta
+        b = self.generate_next_message_bytes(now)
+        cut = random.randrange(len(b))
+        if cut > 0:
+            b = b[cut:]
+        self.pending_json_bytes.extend(b)
+        # Now two interleaved loops:
+        # 1) Generate messages every self.message_delta
+        # 2) Emit a chunk of bytes from pending_json_bytes every self.chunk_delta.
+        while True:
+            if self.stop.is_set():
+                self.logger.info('Returning from ArduinoSimulator.run EXIT')
+                return
+            now = datetime.datetime.now()
+            if now >= self.next_chunk_time:
+                self.output_next_chunk(now)
+            if now >= self.next_message_time:
+                self.generate_next_message(now)
+            self.read_and_handle_relay_queue()
+            if self.pending_json_bytes and self.next_chunk_time < self.next_message_time:
+                next_time = self.next_chunk_time
+            else:
+                next_time = self.next_message_time
+            self.read_relay_queue_until(next_time)
+
+    def read_and_handle_relay_queue(self):
+        """Read all available entries in the relay queue, then handle any complete commands."""
+        while not self.relay_queue.empty():
+            b = self.relay_queue.get_nowait()
+            assert isinstance(b, (bytes, bytearray))
+            self.pending_relay_bytes.extend(b)
+        self.handle_pending_relay_bytes()
+
+    def handle_pending_relay_bytes(self):
+        """Process complete relay commands. NOT YET IMPLEMENTED."""
+        if len(self.pending_relay_bytes):
+            self.logger.error('ArduinoSimulator.handle_pending_relay_bytes NOT YET IMPLEMENTED')
+            self.logger.error('Dropping bytes: {!r}', self.pending_relay_bytes)
+            self.pending_relay_bytes.clear()
+        pass
+
+    def read_relay_queue_until(self, next_time):
+        """Wait for a little bit for relay input, not more than the time until the next action."""
+        now = datetime.datetime.now()
+        if now >= next_time:
+            # Already reached the time for the next main loop event,
+            # so return to repeat the main loop.
+            return
+        remaining = (next_time - now).total_seconds()
+        assert remaining > 0
```

Comment:
It doesn't look like you are catching this assert anywhere which means if it every happens you will get a runtime error.

Reply:
But that makes it a failure in a test, not prod.

Reply:
Ok, shouldn't be an issue then.